File: tests/perl/src/T004_GmsecException.pl
#!/usr/bin/env perl

use strict;
use Test;
use libgmsec_perl;


sub Test_GmsecException
{
	my ($test) = @_;

	$test->setDescription("Test GmsecException");

	test_constructor_1($test);
	test_constructor_2($test);
	test_constructor_3($test);
	test_copy_constructor($test);
	# Note: Accessors are tested in each test above.
}


sub test_constructor_1
{
	libgmsec_perl::logInfo("Test constructor (3 args)");

	my ($test) = @_;

	my $emsg1 = "Something bad happened";

	my $e1 = libgmsec_perl::GmsecException->new(1, 2, $emsg1);

	$test->check("Unexpected error classification", $e1->getErrorClass() == 1);
	$test->check("Unexpected error code", $e1->getErrorCode() == 2);
	$test->check("Unexpected custom code", $e1->getCustomCode() == 0);
	$test->check("Unexpected error message", $e1->getErrorMessage() eq $emsg1);
	$test->check("Unexpected exception message", $e1->what() eq "[1,2,0] : $emsg1");
}


sub test_constructor_2
{
	libgmsec_perl::logInfo("Test constructor (4 args)");

	my ($test) = @_;

	my $emsg1 = "Something bad happened";

	my $e1 = libgmsec_perl::GmsecException->new(1, 2, 3, $emsg1);

	$test->check("Unexpected error classification", $e1->getErrorClass() == 1);
	$test->check("Unexpected error code", $e1->getErrorCode() == 2);
	$test->check("Unexpected custom code", $e1->getCustomCode() == 3);
	$test->check("Unexpected error message", $e1->getErrorMessage() eq $emsg1);
	$test->check("Unexpected exception message", $e1->what() eq "[1,2,3] : $emsg1");
}


sub test_constructor_3
{
	libgmsec_perl::logInfo("Test constructor (Status arg)");

	my ($test) = @_;

	my $reason = "Some worthy status";
	my $status = libgmsec_perl::Status->new(1, 2, $reason, 3);

	my $e1 = libgmsec_perl::GmsecException->new($status);

	$test->check("Unexpected error classification", $e1->getErrorClass() == 1);
	$test->check("Unexpected error code", $e1->getErrorCode() == 2);
	$test->check("Unexpected custom code", $e1->getCustomCode() == 3);
	$test->check("Unexpected error message", $e1->getErrorMessage() eq $reason);
	$test->check("Unexpected exception message", $e1->what() eq "[1,2,3] : $reason");
}


sub test_copy_constructor
{
	libgmsec_perl::logInfo("Test copy-constructor");

	my ($test) = @_;

	my $emsg = "Some worthy exception";

	my $e1 = libgmsec_perl::GmsecException->new(1, 2, 3, $emsg);
	my $e2 = libgmsec_perl::GmsecException->new($e1);

	$test->check("Unexpected error classification", $e1->getErrorClass() == $e2->getErrorClass());
	$test->check("Unexpected error code", $e1->getErrorCode() == $e2->getErrorCode());
	$test->check("Unexpected custom code", $e1->getCustomCode() == $e2->getCustomCode());
	$test->check("Unexpected error message", $e1->getErrorMessage() eq $e2->getErrorMessage());
	$test->check("Unexpected exception message", $e1->what eq $e2->what());
}


Test::driver(\&Test_GmsecException);

File: tests/ruby/src/T004_GmsecException.rb
#!/usr/bin/env ruby

require 'libgmsec_ruby'

load 'Test.rb'


class Test_GmsecException < Test

	def initialize(args)
		super(args)
	end


    def test()
        test_constructor_1
        test_constructor_2
        test_constructor_3
        test_copy_constructor
        # Note: Accessors are tested in each test above.
	end


    def test_constructor_1()
        Libgmsec_ruby::Log::info("Test constructor (3 args)")

        emsg1 = "Something bad happened"

        e1 = Libgmsec_ruby::GmsecException.new(1, 2, emsg1)

        check("Unexpected error classification", e1.get_error_class() == 1)
        check("Unexpected error code", e1.get_error_code() == 2)
        check("Unexpected custom code", e1.get_custom_code() == 0)
        check("Unexpected error message", e1.get_error_message() == emsg1)
        check("Unexpected exception message", e1.message == "[1,2,0] : " + emsg1)
	end


    def test_constructor_2()
        Libgmsec_ruby::Log::info("Test constructor (4 args)")

        emsg1 = "Something bad happened"

        e1 = Libgmsec_ruby::GmsecException.new(1, 2, 3, emsg1)

        check("Unexpected error classification", e1.get_error_class() == 1)
        check("Unexpected error code", e1.get_error_code() == 2)
        check("Unexpected custom code", e1.get_custom_code() == 3)
        check("Unexpected error message", e1.get_error_message() == emsg1)
        check("Unexpected exception message", e1.message == "[1,2,3] : " + emsg1)
	end


    def test_constructor_3()
        Libgmsec_ruby::Log::info("Test constructor (Status arg)")

        reason = "Some worthy status"
        status = Libgmsec_ruby::Status.new(1, 2, reason, 3)

        e1 = Libgmsec_ruby::GmsecException.new(status)

        check("Unexpected error classification", e1.get_error_class() == 1)
        check("Unexpected error code", e1.get_error_code() == 2)
        check("Unexpected custom code", e1.get_custom_code() == 3)
        check("Unexpected error message", e1.get_error_message() == reason)
        check("Unexpected exception message", e1.message == "[1,2,3] : " + reason)
	end


    def test_copy_constructor()
        Libgmsec_ruby::Log::info("Test copy-constructor")

        emsg = "Some worthy exception"

        e1 = Libgmsec_ruby::GmsecException.new(1, 2, 3, emsg)
        e2 = Libgmsec_ruby::GmsecException.new(e1)

        check("Unexpected error classification", e1.get_error_class() == e2.get_error_class())
        check("Unexpected error code", e1.get_error_code() == e2.get_error_code())
        check("Unexpected custom code", e1.get_custom_code() == e2.get_custom_code())
        check("Unexpected error message", e1.get_error_message() == e2.get_error_message())
        check("Unexpected exception message", e1.message == e2.message)
	end
end


begin
    test = Test_GmsecException.new( ARGV )
    test.run
end
